api/v1alpha1/gateway_types.go

```diff
@@ -54,78 +57,185 @@ type GatewayList struct {
+	// connections.
+	//
+	// +optional
 	Listeners []Listener `json:"listeners" protobuf:"bytes,2,rep,name=listeners"`
-	// Routes specifies a schema for associating routes with the Gateway using
-	// selectors. A route is a resource capable of servicing a request and allows
-	// a cluster operator to expose a cluster resource (i.e. Service) by
-	// externally-reachable URL, load-balance traffic and terminate SSL/TLS.
-	// Typically, a route is a "httproute" or "tcproute" in group
-	// "networking.x-k8s.io". However, an implementation may support other resources.
+
+	// Address requested for this gateway. This is optional and behavior
+	// can depend on the GatewayClass. If a value is set in the spec and
+	// the requested address is invalid, the GatewayClass MUST indicate
+	// this in the associated entry in GatewayStatus.Listeners. If no
+	// addresses are specified, the GatewayClass may schedule the Gateway
+	// in an implementation-defined manner, providing it with an appropriate
+	// set of addresses.
 	//
-	// Support: Core
+	// Support:
 	//
-	Routes RouteBindingSelector `json:"routes" protobuf:"bytes,3,opt,name=routes"`
+	// +optional
+	Addresses []ListenerAddress `json:"endpoints" protobuf:"bytes,3,rep,name=endpoints"`
 }
 
+// ProtocolType defines the application protocol accepted by a Listener.
+type ProtocolType string
+
 const (
-	// HTTPProtocol constant.
-	HTTPProtocol = "HTTP"
-	// HTTPSProtocol constant.
-	HTTPSProtocol = "HTTPS"
+	// HTTPProtocolType accepts cleartext HTTP/1.1 sessions over TCP.
+	HTTPProtocolType ProtocolType = "HTTP"
+
+	// HTTPSProtocolType accepts HTTP/1.1 or HTTP/2 sessions over TLS.
+	HTTPSProtocolType ProtocolType = "HTTPS"
+
+	// TLSProtocolType accepts TLS sessions over TCP.
+	TLSProtocolType ProtocolType = "TLS"
+
+	// TCPProtocolType accepts TCP sessions.
+	TCPProtocolType ProtocolType = "TCP"
 )
 
-// Listener defines a
+// HostnameMatchType specifies the types of matches that are valid
+// for host names.
+type HostnameMatchType string
+
+const (
+	// HostnameMatchDomain specifies that the host name provided by the
+	// client should be matched against a DNS domain value. The domain
+	// match should follow the wildcard certificate matching rules in
+	// [RFC 6125](https://tools.ietf.org/html/rfc6125#section-6.4.3).
+	HostnameMatchDomain HostnameMatchType = "Domain"
+
+	// HostnameMatchExact specifies that the host name provided by the
+	// client must exactly match the specified value.
+	HostnameMatchExact HostnameMatchType = "Exact"
+)
+
+// HostnameMatch specifies how a Listener should match the host name from a
+// client request. Depending on the incoming protocol, the match may apply to
+// the TLS and the HTTP protocol.
+type HostnameMatch struct {
+	// Match specifies how the host name provided by the client should be
+	// matched against the given value.
+	//
+	// +optional
+	// +kubebuilder:validation:Enum=Domain;Exact
+	// +kubebuilder:default=Exact
```

Comment:
The default of `Exact` seems ambiguous. My understanding here is that if the value of the match is is empty, that implies a match against `*`, meaning all hosts are accepted.

Should the default then be:
```yaml
match: domain
value: "*"
```

Reply:
The idea is that the `Hostname` field specifies a match and that the `Name` field in `HostnameMatch` must not be empty. The way to match all traffic (i.e. route to all hosts) is to omit the `Hostname` field altogether. I'll update the language to try and clarify this.

Reply:
Looks good now, resolving this thread.

Reply:
> I think it's clearer in an enumeration to have an explicit HostnameMatchAny and make that the default value, instead of empty

+1. This seems more readable and user-friendly.